x/evm/keeper/keeper.go

```diff
@@ -213,166 +174,295 @@ func (k Keeper) GetTxIndexTransient() uint64 {
+func (k *Keeper) PostTxProcessing(ctx sdk.Context, from common.Address, to *common.Address, receipt *ethtypes.Receipt) error {
+	if k.hooks == nil {
+		return nil
 	}
+	return k.hooks.PostTxProcessing(ctx, from, to, receipt)
+}
 
-	return logs
+// Tracer return a default vm.Tracer based on current keeper state
+func (k Keeper) Tracer(ctx sdk.Context, msg core.Message, ethCfg *params.ChainConfig) vm.Tracer {
+	return types.NewTracer(k.tracer, msg, ethCfg, ctx.BlockHeight())
 }
 
-// SetLog sets the log for a transaction in the KVStore.
-func (k Keeper) AddLogTransient(log *ethtypes.Log) {
-	store := prefix.NewStore(k.Ctx().TransientStore(k.transientKey), types.KeyPrefixTransientTxLogs)
+// GetAccountWithoutBalance load nonce and codehash without balance,
+// more efficient in cases where balance is not needed.
+func (k *Keeper) GetAccountWithoutBalance(ctx sdk.Context, addr common.Address) (*statedb.Account, error) {
+	cosmosAddr := sdk.AccAddress(addr.Bytes())
+	acct := k.accountKeeper.GetAccount(ctx, cosmosAddr)
+	if acct == nil {
+		return nil, nil
+	}
 
-	key := log.TxHash.Bytes()
-	key = append(key, sdk.Uint64ToBigEndian(uint64(log.Index))...)
+	ethAcct, ok := acct.(*ethermint.EthAccount)
+	if !ok {
+		return nil, errors.New("not EthAccount")
+	}
 
-	txIndexLog := types.NewLogFromEth(log)
-	bz := k.cdc.MustMarshal(txIndexLog)
-	store.Set(key, bz)
+	return &statedb.Account{
+		Nonce:    ethAcct.Sequence,
+		CodeHash: common.FromHex(ethAcct.CodeHash),
+	}, nil
 }
 
-// GetLogSizeTransient returns EVM log index on the current block.
-func (k Keeper) GetLogSizeTransient() uint64 {
-	store := k.Ctx().TransientStore(k.transientKey)
-	bz := store.Get(types.KeyPrefixTransientLogSize)
-	if len(bz) == 0 {
+// GetNonce returns the sequence number of an account, returns 0 if not exists.
+func (k *Keeper) GetNonce(ctx sdk.Context, addr common.Address) uint64 {
+	cosmosAddr := sdk.AccAddress(addr.Bytes())
+	acct := k.accountKeeper.GetAccount(ctx, cosmosAddr)
+	if acct == nil {
 		return 0
 	}
 
-	return sdk.BigEndianToUint64(bz)
+	ethAcct, ok := acct.(*ethermint.EthAccount)
+	if !ok {
+		return 0
+	}
+
+	return ethAcct.Sequence
 }
 
-// IncreaseLogSizeTransient fetches the current EVM log index from the transient store, increases its
-// value by one and then sets the new index back to the transient store.
-func (k Keeper) IncreaseLogSizeTransient() {
-	logSize := k.GetLogSizeTransient()
-	store := k.Ctx().TransientStore(k.transientKey)
-	store.Set(types.KeyPrefixTransientLogSize, sdk.Uint64ToBigEndian(logSize+1))
+// GetBalance load account's balance of gas token
+func (k *Keeper) GetBalance(ctx sdk.Context, addr common.Address) *big.Int {
+	cosmosAddr := sdk.AccAddress(addr.Bytes())
+	params := k.GetParams(ctx)
+	coin := k.bankKeeper.GetBalance(ctx, cosmosAddr, params.EvmDenom)
+	return coin.Amount.BigInt()
 }
 
 // ----------------------------------------------------------------------------
-// Storage
+// statedb.Keeper implementation
```

Comment:
nit, can we move these functions to `x/evm/keeper/statedb.go`?

Reply:
done